File: apps/api/services/mime_allowlist.py
"""File MIME type allowlist (cycle II2, Python half).

Server-side mirror of `apps/web/lib/mime-allowlist.ts`. Used by:

  * The upload endpoint validator (rejects malformed MIME with
    HTTP 422 before the file lands in storage).
  * The avatar uploader API.
  * The audit-evidence attach endpoint.

  is_allowed_mime(type, category)  — bool
  accepted_extensions(category)    — tuple[str, ...]
  MIME_CATEGORIES                  — closed tuple
  MIME_ALLOWLIST                   — frozen category → frozenset

Pinned defenses:
  * SVG explicitly NOT in `photo` (XSS via embedded `<script>`).
  * `application/octet-stream` rejected for ALL categories.
  * HEIC accepted in `photo` (iPhone uploads dominant in VN).

Pure stdlib.
"""

from __future__ import annotations

from typing import Literal

MimeCategory = Literal["photo", "document", "cad", "archive"]


MIME_CATEGORIES: tuple[MimeCategory, ...] = (
    "photo",
    "document",
    "cad",
    "archive",
)


# Closed allowlist. All entries lowercased. Adding a new MIME
# requires touching this AND the TS half AND adding the
# extension to _ACCEPTED_EXTENSIONS — three-way pin enforced
# by tests.
MIME_ALLOWLIST: dict[MimeCategory, frozenset[str]] = {
    "photo": frozenset(
        {
            "image/jpeg",
            "image/png",
            "image/heic",
            "image/webp",
        }
    ),
    "document": frozenset(
        {
            "application/pdf",
        }
    ),
    "cad": frozenset(
        {
            "application/acad",
            "image/vnd.dwg",
            "application/vnd.dxf",
        }
    ),
    "archive": frozenset(
        {
            "application/zip",
        }
    ),
}


_ACCEPTED_EXTENSIONS: dict[MimeCategory, tuple[str, ...]] = {
    "photo": (".jpg", ".jpeg", ".png", ".heic", ".webp"),
    "document": (".pdf",),
    "cad": (".dwg", ".dxf"),
    "archive": (".zip",),
}


def is_allowed_mime(
    mime_type: str | None,
    category: MimeCategory,
) -> bool:
    """True iff `mime_type` is in the allowlist for `category`.

    Defensive normalisation:
      * Lowercased on comparison.
      * Whitespace stripped.
      * MIME parameters stripped (`image/jpeg; charset=binary`
        → `image/jpeg`).
      * Empty / None → False.
    """
    if not mime_type:
        return False
    lower = mime_type.lower().strip()
    base_type = lower.split(";")[0].strip()
    return base_type in MIME_ALLOWLIST.get(category, frozenset())


def accepted_extensions(category: MimeCategory) -> tuple[str, ...]:
    """File extensions accepted by `category`. Used by the API
    `accept=` hint emitted in upload-config endpoints."""
    return _ACCEPTED_EXTENSIONS.get(category, ())

File: apps/web/lib/mime-allowlist.ts
/**
 * File MIME type allowlist (cycle II2, TS half).
 *
 * Closed allowlist of MIME types per upload category. Today the
 * upload widget, the avatar uploader, and the audit-evidence
 * attach button each duplicate the type list inline with subtly
 * different MIME entries. This module is the single source of truth.
 *
 *   isAllowedMime(type, category)   — bool
 *   acceptedExtensions(category)    — for input `accept=` attribute
 *   MIME_CATEGORIES                 — closed list
 *   MIME_ALLOWLIST                  — frozen category → set of MIMEs
 *
 * Categories:
 *   * photo    — image/jpeg, image/png, image/heic, image/webp
 *   * document — application/pdf
 *   * cad      — application/acad, image/vnd.dwg, application/vnd.dxf
 *   * archive  — application/zip
 *
 * Pinned defenses:
 *   * SVG explicitly NOT in `photo` — the `image/svg+xml` MIME
 *     allows embedded `<script>` (XSS vector). Pin so a refactor
 *     that re-adds it surfaces in review.
 *   * `application/octet-stream` rejected for ALL categories —
 *     type-confusion guard against malicious uploads with a
 *     spoofed MIME.
 *   * HEIC accepted in `photo` — iPhone uploads dominant in VN.
 *
 * Pure TS. Mirrors `apps/api/services/mime_allowlist.py`.
 */


export type MimeCategory = "photo" | "document" | "cad" | "archive";


/** Closed list of upload categories. Pin via test. */
export const MIME_CATEGORIES: readonly MimeCategory[] = [
  "photo",
  "document",
  "cad",
  "archive",
];


/** Closed allowlist: category → set of accepted MIME types
 *  (lowercased). All comparisons happen against the lowercased
 *  base type (parameters stripped). */
export const MIME_ALLOWLIST: Readonly<Record<MimeCategory, ReadonlySet<string>>> = {
  photo: new Set([
    "image/jpeg",
    "image/png",
    "image/heic",
    "image/webp",
  ]),
  document: new Set([
    "application/pdf",
  ]),
  cad: new Set([
    "application/acad",
    "image/vnd.dwg",
    "application/vnd.dxf",
  ]),
  archive: new Set([
    "application/zip",
  ]),
};


/** Accepted file extensions per category. Used for the upload
 *  widget's `accept=` attribute (which accepts both MIME types
 *  and extensions). */
const _ACCEPTED_EXTENSIONS: Readonly<Record<MimeCategory, readonly string[]>> = {
  photo: [".jpg", ".jpeg", ".png", ".heic", ".webp"],
  document: [".pdf"],
  cad: [".dwg", ".dxf"],
  archive: [".zip"],
};


/**
 * True iff the MIME type is in the allowlist for `category`.
 *
 * Defensive normalisation:
 *   * Lowercased on comparison.
 *   * Whitespace stripped.
 *   * MIME parameters stripped (`image/jpeg; charset=binary` →
 *     `image/jpeg`).
 *   * Empty / null / undefined → false.
 */
export function isAllowedMime(
  mimeType: string | null | undefined,
  category: MimeCategory,
): boolean {
  if (!mimeType) return false;
  const lower = mimeType.toLowerCase().trim();
  // Strip parameters: take only the part before any `;`.
  const baseType = (lower.split(";")[0] ?? "").trim();
  return MIME_ALLOWLIST[category].has(baseType);
}


/** File extensions accepted by `category`. Used as the
 *  `<input type="file" accept="...">` attribute value. */
export function acceptedExtensions(
  category: MimeCategory,
): readonly string[] {
  return _ACCEPTED_EXTENSIONS[category];
}
